SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp

```diff
@@ -0,0 +1,360 @@
+	return false;
+}
+
+bool USpatialLatencyTracer::EndLatencyTrace(UObject* WorldContextObject)
+{
+#if TRACE_LIB_ACTIVE
+	if (USpatialLatencyTracer* Tracer = GetTracer(WorldContextObject))
+	{
+		return Tracer->EndLatencyTrace_Internal();
+	}
+#endif // TRACE_LIB_ACTIVE
+	return false;
+}
+
+#if TRACE_LIB_ACTIVE
+bool USpatialLatencyTracer::IsValidKey(const TraceKey Key)
+{
+	FScopeLock Lock(&Mutex);
+	return TraceMap.Find(Key);
+}
+
+TraceKey USpatialLatencyTracer::GetTraceKey(const UObject* Obj, const UFunction* Function)
+{
+	FScopeLock Lock(&Mutex);
+
+	ActorFuncKey FuncKey{ Cast<AActor>(Obj), Function };
+	TraceKey ReturnKey = InvalidTraceKey;
+	TrackingTraces.RemoveAndCopyValue(FuncKey, ReturnKey);
+	return ReturnKey;
+}
+
+void USpatialLatencyTracer::WriteToLatencyTrace(const TraceKey Key, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	if (TraceSpan* Trace = TraceMap.Find(Key))
+	{
+		WriteKeyFrameToTrace(Trace, TraceDesc);
+	}
+}
+
+void USpatialLatencyTracer::EndLatencyTrace(const TraceKey Key, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	if (TraceSpan* Trace = TraceMap.Find(Key))
+	{
+		WriteKeyFrameToTrace(Trace, TraceDesc);
+
+		Trace->End();
+		TraceMap.Remove(Key);
+	}
+}
+
+void USpatialLatencyTracer::WriteTraceToSchemaObject(const TraceKey Key, Schema_Object* Obj, const Schema_FieldId FieldId)
+{
+	FScopeLock Lock(&Mutex);
+
+	if (TraceSpan* Trace = TraceMap.Find(Key))
+	{
+		Schema_Object* TraceObj = Schema_AddObject(Obj, FieldId);
+
+		const improbable::trace::SpanContext& TraceContext = Trace->context();
+		improbable::trace::TraceId _TraceId = TraceContext.trace_id();
+		improbable::trace::SpanId _SpanId = TraceContext.span_id();
+
+		SpatialGDK::AddBytesToSchema(TraceObj, SpatialConstants::UNREAL_RPC_TRACE_ID, &_TraceId[0], _TraceId.size());
+		SpatialGDK::AddBytesToSchema(TraceObj, SpatialConstants::UNREAL_RPC_SPAN_ID, &_SpanId[0], _SpanId.size());
+	}
+}
+
+TraceKey USpatialLatencyTracer::ReadTraceFromSchemaObject(Schema_Object* Obj, const Schema_FieldId FieldId)
+{
+	FScopeLock Lock(&Mutex);
+
+	check(GetActiveTrace() == nullptr);
+
+	if (Schema_GetObjectCount(Obj, FieldId) > 0)
+	{
+		Schema_Object* TraceData = Schema_IndexObject(Obj, FieldId, 0);
+
+		const uint8* TraceBytes = Schema_GetBytes(TraceData, SpatialConstants::UNREAL_RPC_TRACE_ID);
+		const uint8* SpanBytes = Schema_GetBytes(TraceData, SpatialConstants::UNREAL_RPC_SPAN_ID);
+
+		improbable::trace::TraceId _TraceId;
+		memcpy(&_TraceId[0], TraceBytes, sizeof(improbable::trace::TraceId));
+
+		improbable::trace::SpanId _SpanId;
+		memcpy(&_SpanId[0], SpanBytes, sizeof(improbable::trace::SpanId));
+
+		improbable::trace::SpanContext DestContext(_TraceId, _SpanId);
+
+		FString SpanMsg = FormatMessage(TEXT("Read Trace From Schema Obj"));
+		TraceSpan RetrieveTrace = improbable::trace::Span::StartSpanWithRemoteParent(TCHAR_TO_UTF8(*SpanMsg), DestContext);
+		TraceMap.Add(ActiveTraceKey, MoveTemp(RetrieveTrace));
```

Comment:
Does this rely on us not receiving another trace while processing this one? Does this get handled when we queue up RPCs for future processing or have I just misunderstood?

Reply:
Or do we only read this once we know we can process the RPC?

Reply:
We process this when we're about to apply the RPC.